Fix issues #43183 & #42585: opensearch exporter dynamic log index feature puts logs in wrong index & support multiple placeholders

Confushion opened the pull request:
#### Description
- Modified the OpenSearch exporter to resolve index names per span/log record instead of per resource, using the specified attribute precedence (item > scope > resource). 
The changes ensure that each individual span or log record can be indexed into a different index based on its attributes, while maintaining backward compatibility for default configurations.

- Added support for multiple placeholders in index names.
This enhances flexibility for dynamic indexing based on multiple attributes, enabling more granular index naming strategies.

#### Link to tracking issue
- Fixes #43183
- Fixes #42585

#### Testing
- Update tests to use new per-item methods and verify precedence
- Added tests for multiple placeholder resolution and fallback behavior


#### Documentation
-  Update documentation with examples of multiple placeholder usage

mgangelen committed:
Push indexName resolving one level lower in the loop

mgangelen committed:
Remove duplicate tests, already done in file index_resolver_test.go

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

MovieStoreGuy commented on exporter/opensearchexporter/sso_log_exporter_test.go:
Any reason why this entire file was deleted?

Confushion replied:
That file only contained tests for index name resolving. It was renamed to index_resolver_test.go as the functions being tested are in file index_resolver.go

Confushion committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

KarstenSchnitter commented on exporter/opensearchexporter/index_resolver_test.go:
```diff
@@ -193,13 +193,73 @@ func TestIndexResolver_ResolveTraceIndex_UnknownPlaceholder(t *testing.T) {
+func TestIndexResolver_ResolveLogIndex_MultipleResources(t *testing.T) {
+	resolver := newIndexResolver()
+	cfg := &Config{
+		LogsIndex:           "%{service.name}-logs",
```
What if the index name is not a resource attribute but a scope or log/span attribute? In that case, the logic would still fail and select only the last received value, right?

KarstenSchnitter reviewed (commented):
Thanks for providing the change. I like, that it gives more control on the index name. During an OpenSearch bulk insert, the index name is provided for each document separately. However, the new extraction logic only supports index name placeholders based on resource attributes. In principle, all attributes could be enabled for index name placeholders. But this would required more changes to the source code still.

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
feat: resolve index names per span/log record with attribute precedence

- Change index resolution from per-ResourceSpans/ResourceLogs to per-span/log record
- Implement attribute precedence: item attributes > scope attributes > resource attributes
- Add ResolveSpanIndex and ResolveLogRecordIndex methods for per-item resolution
- Add collectSpanAttributes and collectLogRecordAttributes with precedence logic
- Update bulk indexers to resolve index per item in the visitor functions
- Remove unused methods: ResolveTraceIndex, ResolveLogIndex, collectLogAttributes, collectTraceAttributes
- Update tests to use new per-item methods and verify precedence

This enables more granular indexing based on individual span/log attributes, improving flexibility for users.

mgangelen committed:
feat: support multiple placeholders in index names #42585

- Remove validation restricting index patterns to one placeholder
- Allow multiple %{key} placeholders per logs_index and traces_index
- Update documentation with examples of multiple placeholder usage
- Add tests for multiple placeholder resolution and fallback behavior

This enhances flexibility for dynamic indexing based on multiple attributes, enabling more granular index naming strategies.

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

KarstenSchnitter reviewed (approved):
This looks good to me. Just a recommendation to warn people about cardinality of the attributes being used. They may risk destabilizing the OpenSearch cluster by lots of custom index names.

atoulme commented:
@ps48 please review as codeowner

mgangelen committed:
Implement missing tests

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Added Changelog Entries

mgangelen committed:
Merge branch 'Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-' of github.com:Confushion/opentelemetry-collector-contrib into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

ps48 commented on exporter/opensearchexporter/README.md:
```diff
@@ -47,10 +47,10 @@ The OpenSearch exporter supports dynamic index names for both logs and traces us
-  - Example: `otel-logs-%{service.name}` or `otel-traces-%{custom.label}`
-  - The value is looked up from resource attributes, log/span attributes, and scope attributes
+  - Example: `otel-logs-%{service.name}` or `otel-traces-%{service.name}-%{env}`
+  - Multiple placeholders are supported per index name
```
Should we cap the max placeholders? Will not adding a max value cause an unintended issues?

Confushion replied:
My opinion is that capping this option would limit the user too much. It is up to the user to check the limits of his setup. There could be use cases where a user has the need for more placeholders, the user should make sure his infra id up for the job.

ps48 reviewed (commented):
This is a great contribution thanks @Confushion! Have some questions/suggestions for some of the changes. 

cc: @atoulme @MovieStoreGuy @songy23

Confushion commented on exporter/opensearchexporter/log_bulk_indexer.go:
```diff
@@ -58,22 +58,23 @@ func (lbi *logBulkIndexer) appendRetryLogError(err error, log plog.Logs) {
-		payload, err := lbi.model.encodeLog(resource, scope, scopeSchemaURL, log)
+func (lbi *logBulkIndexer) submit(ctx context.Context, ld plog.Logs, ir *indexResolver, cfg *Config, timestamp time.Time) {
+	forEachLog(ld, ir, cfg, timestamp, func(resolver *indexResolver, cfg *Config, timestamp time.Time, resource pcommon.Resource, resourceSchemaURL string, scope pcommon.InstrumentationScope, scopeSchemaURL string, logRecord plog.LogRecord) {
+		indexName := resolver.ResolveLogRecordIndex(cfg, resource, scope, logRecord, timestamp)
```
That seems like a good improvement, is it ok that I create a new improvement PR for this? It is not actually related to the current bug and I would like to have the bugfix out so that I can actually use the opensearch exporter in my deployment…

Confushion commented on exporter/opensearchexporter/index_resolver.go:
```diff
@@ -73,84 +49,85 @@ func (*indexResolver) appendTimeFormat(index, timeFormat string, timestamp time.
-		}
+	// Resource attributes (lowest precedence)
+	resAttrs := resource.Attributes()
+	resAttrs.Range(func(k string, v pcommon.Value) bool {
```
That seems like a good improvement, is it ok that I create a new improvement PR for this? It is not actually related to the current bug and I would like to have the bugfix out so that I can actually use the opensearch exporter in my deployment…

ps48 commented on exporter/opensearchexporter/index_resolver_test.go:
```diff
@@ -193,13 +220,222 @@ func TestIndexResolver_ResolveTraceIndex_UnknownPlaceholder(t *testing.T) {
+		t.Errorf("expected %q, got %q", expected, index)
+	}
+}
+
```
Thanks! will wait for the test to be added.

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Implement tests for attribute precedence

mgangelen committed:
Restrict attribute mapping to placeholder keys only

mgangelen committed:
Cache attributeMaps as much as possible and simplify index_resolver and index_resolver_test

Confushion committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Implement tests for *_bulk_indexer

mgangelen committed:
Merge branch 'Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-' of github.com:Confushion/opentelemetry-collector-contrib into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Optimize timeSuffix calculation

Confushion commented:
> This is a great contribution thanks @Confushion! Have some questions/suggestions for some of the changes.
> 
> cc: @atoulme @MovieStoreGuy @songy23

All suggestions have been implemented.

ps48 commented:
@Confushion thanks for responding to all the questions and implementing the suggestions 🙏🏽 This is really a great feature contribution for the OS exporter.
@MovieStoreGuy @songy23 can you please help run the workflows and merge this PR.

mgangelen committed:
Rewrite if-else to switch statement (Lint requirement)

mgangelen committed:
Remove unused parameter 'cfg' (Lint requirement)

Confushion commented:
Fixed workflow Lint issues.

ChrsMark committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

ChrsMark commented:
Approved by code-owners and @MovieStoreGuy already. It can be merged once CI passes.

ChrsMark commented:
@Confushion it seems there are still some lint errors.

mgangelen committed:
Remove unused parameter 'timestap' (Lint requirement)

mgangelen committed:
Readability: don't use inline functions as parameters

mgangelen committed:
Merge branch 'Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-' of github.com:Confushion/opentelemetry-collector-contrib into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'open-telemetry:main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Fix integration tests by re-adding default index name for backwards compatibility

mgangelen committed:
remove obsolete testdata configs

mgangelen committed:
Merge branch 'Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-' of github.com:Confushion/opentelemetry-collector-contrib into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

Confushion committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

ChrsMark committed:
Merge branch 'main' into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-

mgangelen committed:
Update external link

mgangelen committed:
Merge branch 'Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-' of github.com:Confushion/opentelemetry-collector-contrib into Fix-issue-#43183-OpensearchExporter-Dynamic-log-index-feature-puts-logs-in-wrong-index-
